pkg/component/controller/workerconfig/cleaner.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2022 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workerconfig
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+
+	"github.com/k0sproject/k0s/pkg/applier"
+	"github.com/k0sproject/k0s/pkg/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	"go.uber.org/multierr"
+)
+
+type cleaner interface {
+	init()
+	reconciled(context.Context)
+	stop()
+}
+
+// kubeletConfigCleaner takes care of removing the bits
+// and pieces of the replaced kubelet-config component.
+type kubeletConfigCleaner struct {
+	log logrus.FieldLogger
+
+	dir           string
+	clientFactory kubernetes.ClientFactoryInterface
+
+	doStop atomic.Value
+}
+
+func (c *kubeletConfigCleaner) init() {
+	var errs []error
+	for _, path := range []string{filepath.Join(c.dir, "kubelet-config.yaml"), c.dir} {
+		if err := os.Remove(path); err == nil {
+			c.log.Debugf("Removed deprecated path: %q", path)
+		} else if os.IsNotExist(err) {
+			c.log.Debugf("Deprecated path doesn't exist: %q", path)
+		} else {
+			errs = append(errs, err)
+		}
+	}
+
+	if err := multierr.Combine(errs...); err != nil {
+		c.log.WithError(multierr.Combine(errs...)).Warn("Failed to remove some deprecated paths")
+	}
+}
+
+func (c *kubeletConfigCleaner) reconciled(ctx context.Context) {
+	ctx, cancel := context.WithCancel(ctx)
+	done := make(chan struct{})
+
+	// Execute this only once ...
+	if !c.doStop.CompareAndSwap(nil, func() {
```

Comment:
I would recommend to use https://pkg.go.dev/sync instead, since it does exactly what is required and it is a part of std library.


```go
c.Once.Do(func() {...})

Reply:
Yes, sync.Once makes it more straight forward, but it doesn't have anything built-in for async cancellation. How would the stop call cancel a concurrent reconciliation then? Do you have a suggestion?

```diff
diff --git a/pkg/component/controller/workerconfig/cleaner.go b/pkg/component/controller/workerconfig/cleaner.go
--- a/pkg/component/controller/workerconfig/cleaner.go
+++ b/pkg/component/controller/workerconfig/cleaner.go
@@ -20,7 +20,7 @@ import (
 	"context"
 	"os"
 	"path/filepath"
-	"sync/atomic"
+	"sync"
 
 	"github.com/k0sproject/k0s/pkg/applier"
 	"github.com/k0sproject/k0s/pkg/kubernetes"
@@ -43,7 +43,7 @@ type kubeletConfigCleaner struct {
 	dir           string
 	clientFactory kubernetes.ClientFactoryInterface
 
-	doStop atomic.Value
+	reconcileOnce sync.Once
 }
 
 func (c *kubeletConfigCleaner) init() {
@@ -64,34 +64,18 @@ func (c *kubeletConfigCleaner) init() {
 }
 
 func (c *kubeletConfigCleaner) reconciled(ctx context.Context) {
-	ctx, cancel := context.WithCancel(ctx)
-	done := make(chan struct{})
-
-	// Execute this only once ...
-	if !c.doStop.CompareAndSwap(nil, func() {
-		cancel()
-		<-done
-	}) {
-		return
-	}
-
-	go func() {
-		defer func() {
-			close(done)
-			c.stop()
-		}()
-
+	go c.reconcileOnce.Do(func() {
 		if err := c.removeKubeletStack(ctx); err != nil {
 			c.log.WithError(err).Warn("Failed to remove deprecated kubelet stack")
 		}
-	}()
+	})
 }
 
 func (c *kubeletConfigCleaner) stop() {
-	if stop, ok := c.doStop.Swap(func() {}).(func()); ok {
-		stop()
-		return
-	}
+	c.reconcileOnce.Do(func() {
+		// This call either blocks until a concurrent reconciliation has
+		// finished, or prevents it from starting.
+	})
 }
 
 func (c *kubeletConfigCleaner) removeKubeletStack(ctx context.Context) error {
```

Reply:
What is happening here exactly? 
I don't fully follow, what the code does. It looks like there is no need for either `atomic.Value` and  `sync.Once`

Reply:
This code was intended to remove the deprecated stack from the cluster. As this is rather long-running, the intent was to run it once in a goroutine that should be "cancellable" and "awaitable" for completion.

However, the better and much easier approach is not to remove the old stack, but add some deprecation notes to it. Otherwise, old 1.25 workers wouldn't work with a 1.26 control plane. The actual removal of the stack can be done in 1.27, by simply removing (or renaming) all `*.yaml` files in the `manifests/kubelet` directory.

So all of this is gone by now.